xarray/core/computation.py

```diff
@@ -579,28 +589,41 @@ def apply_variable_ufunc(
                 "``.load()`` or ``.compute()``"
             )
         elif dask == "parallelized":
-            input_dims = [broadcast_dims + dims for dims in signature.input_core_dims]
             numpy_func = func
 
+            if dask_gufunc_kwargs is None:
+                dask_gufunc_kwargs = {}
+
             def func(*arrays):
-                return _apply_blockwise(
-                    numpy_func,
-                    arrays,
-                    input_dims,
-                    output_dims,
-                    signature,
-                    output_dtypes,
-                    output_sizes,
-                    meta,
+                import dask.array as da
+
+                res = da.apply_gufunc(
+                    numpy_func, str(signature), *arrays, **dask_gufunc_kwargs,
```

Comment:
I think it's important to use `to_gufunc_string()`:
```suggestion
                    numpy_func, signature.to_gufunc_string(), *arrays, **dask_gufunc_kwargs,
```

Otherwise, apply_ufunc would break when called on arrays with dimension names that are not valid Python identifiers.

Reply:
Hmm, here I would need a bit more assistance, example taken from `test_apply_dask_parallelized_errors()`

```python
import xarray as xr

import dask.array as da

array = da.ones((2, 2), chunks=(1, 1))
data_array = xr.DataArray(array, dims=("x", "y"))

def identity(x):
    return x

xr.apply_ufunc(
        identity,
        data_array,
        dask="parallelized",
        output_dtypes=[float],
        input_core_dims=[("y",)],
        output_core_dims=[("y",)],
    )
```
The error raised using `signature.to_gufunc_string()`

```
ValueError: Core dimension `'dim0'` consists of multiple chunks. To fix, rechunk into a single chunk along this dimension or set `allow_rechunk=True`, but beware that this may increase memory usage significantly.
```

The error raised using `str(signature)`

```
ValueError: Core dimension `'y'` consists of multiple chunks. To fix, rechunk into a single chunk along this dimension or set `allow_rechunk=True`, but beware that this may increase memory usage significantly.
```

For the latter I can create a working test, for the former not. I've no idea how to work around this.

Reply:
At least I would have my problems with the error message, because there is no `dim0` in the source data.

Reply:
@shoyer This test breaks within dask using `signature.to_gufunc_string()`:

```python
    import dask.array as da

    array = da.ones((2, 2), chunks=(1, 1))
    data_array = xr.DataArray(array, dims=("x", "y"))

    def stack_negative(obj):
        def func(x):
            return np.stack([x, -x], axis=-1)

        return apply_ufunc(
            func,
            obj,
            output_core_dims=[["sign"]],
            dask="parallelized",
            output_dtypes=[obj.dtype],
            output_sizes={"sign": 2},
        )
    stack_negative(data_array)
```

```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_computation.py:776: in stack_negative
    return apply_ufunc(
core/computation.py:1020: in apply_ufunc
    return apply_dataarray_vfunc(
core/computation.py:234: in apply_dataarray_vfunc
    result_var = func(*data_vars)
core/computation.py:630: in apply_variable_ufunc
    result_data = func(*input_data)
core/computation.py:600: in func
    res = da.apply_gufunc(
/home/kai/miniconda/envs/wradlib_38_01/lib/python3.8/site-packages/dask/array/gufunc.py:461: in apply_gufunc
    core_output_shape = tuple(core_shapes[d] for d in ocd)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

.0 = <tuple_iterator object at 0x7f46d9d3e910>

>   core_output_shape = tuple(core_shapes[d] for d in ocd)
E   KeyError: 'dim0'

/home/kai/miniconda/envs/wradlib_38_01/lib/python3.8/site-packages/dask/array/gufunc.py:461: KeyError

```

Reply:
How can this be resolved properly?

Reply:
Any ideas here?

Reply:
According to the docstring `to_gufunc_string` "handles dimensions that don't map to Python identifiers". I am honestly not entirely sure what kind of dimension names these are? Non-string? 

Anyway - what it does is to map the dimension names like so: 
```
    "(x),(x,y)->(z)" -> "(dim0),(dim0,dim1)->(dim2)"
```

Now in `stack_negative` the name of `output_sizes` (`"sign"`) and the `signature` (`"()->(dim0)"`) are inconsistent. A possible way forward is:

1. re-map the names in `output_sizes` to the new names
2. Open a new issue to use the original names in `to_gufunc_string` as much as possible and fix this in a follow up PR. (I also find this confusing in an error message). For the `ValueError: Core dimension 'dim0'` I.e. I would leave this for the moment.

(Alternatively just use `str` and also defer 1. to a later PR. The proposed solution for #4130 would require using `to_gufunc_string` (i.e. I would probably do this))

---

To achieve 2, you would need to expose [dims_map](https://github.com/pydata/xarray/blob/a64cf2d5476e7bbda099b34c40b7be1880dbd39a/xarray/core/computation.py#L130) as a property in `_UFuncSignature` and then translate the `output_sizes` like so
```python
output_sizes = {"dim{}".format(signature.dims_map[key]): value for key, value in output_sizes.items()}
```

Reply:
> I am honestly not entirely sure what kind of dimension names these are? Non-string?

In addition to non-string dimension names I think that also means all strings you can't use as a python variable (e.g. starting with a digit or containing a `-` or `.`)

Reply:
> (Alternatively just use `str` and defer 1. to a later PR. The proposed solution for #4130 would require using `to_gufunc_string` (i.e. I would probably do this))
> 
> To achieve 2, you would need to expose [dims_map](https://github.com/pydata/xarray/blob/a64cf2d5476e7bbda099b34c40b7be1880dbd39a/xarray/core/computation.py#L130) as a property in `_UFuncSignature` and then translate the `output_sizes` like so
> 
> ```python
> output_sizes = {"dim{}".format(signature.dims_map[key]): value for key, value in output_sizes.items()}
> ```

@mathause Your solution 2 looks better from my perspective. I'll try to apply to this.

Reply:
adapted to @mathause's suggestion (2)

Reply:
Just came to my mind: it would be good to throw a nice error message `if key not in dims_map`. Or use `dims_map. get(key, key)` and let dask throw the error.